### ISSUE TITLE
Groups use absolute paths

### ISSUE THREAD

**Issue by MaienM** [colszowka/simplecov#419]
It seems that the groups match the full path, instead of just the part relative to the directory it's all being run from. I noticed this when on my jenkins server, all files where included in the Libraries group (with the filter `lib`), because the files were located somewhere deep in `/var/lib/jenkins/`.

Now this may very well be intentional, but it feels very illogical to me.

This is using simplecov v0.10.0, ruby 2.2.3, rails-api 4.2.4, and rspec 3.3.2.

**Comment by bf4** [comment]
What's `SimpleCov.root` ?  where are you configuring it? How are you running it? Where do you 'start' SimpleCov?  What's an example of an expected path that is included and an unexpected one? Thanks

**Comment by MaienM** [comment]
Sorry for the somewhat delayed response.

`Simplecov.root` is 

```
/var/lib/jenkins/jobs/AEB/workspace
```

I am configuring it through a `.simplecov` file, contents are as follows

``` ruby
require 'simplecov-rcov'
SimpleCov.formatters = [
	SimpleCov::Formatter::HTMLFormatter,
	SimpleCov::Formatter::RcovFormatter,
]
SimpleCov.profiles.define 'aeb' do
	@filters = []

	add_filter '/spec/'
	add_filter '/config/'

	add_group 'Controllers' do |src_file|
		src_file.filename['app/controllers'] && !src_file.filename['concern']
	end
	add_group 'Concerns', 'app/controllers/concerns'
	add_group 'Models', 'app/models'
	add_group 'Docs', 'app/docs'
	add_group 'Libraries' do |src_file|
		src_file.filename['lib/middleware'] || src_file.filename['lib/other']
	end
end
SimpleCov.start 'aeb'
```

As you can see, the Libraries group now checks for certain subfolders of lib manually, which is the workaround we're currently using. That line used to be

``` ruby
add_group 'Libraries', 'lib'
```

I'm running it with the RSpec test cases, by placing `require 'simplecov'` as the first thing in `spec/spec_helper.rb`. Rspec is run by executing `bin/bundle exec rake spec`. 

The working directory while running these commands is

```
/var/lib/jenkins/jobs/AEB/workspace
```

Examples of expected paths are:

```
/var/lib/jenkins/jobs/AEB/workspace/lib/other/permission_initializer.rb
/var/lib/jenkins/jobs/AEB/workspace/lib/ext/string.rb
/var/lib/jenkins/jobs/AEB/workspace/lib/ci/run.rb
/var/lib/jenkins/jobs/AEB/workspace/lib/core_extensions/object/send_chain.rb
/var/lib/jenkins/jobs/AEB/workspace/lib/rubocop/cop/style/indent_hash_sensible.rb
```

Expamples of unexpected paths are:

```
/var/lib/jenkins/jobs/AEB/workspace/app/models/concerns/permission_holder.rb
/var/lib/jenkins/jobs/AEB/workspace/app/models/category.rb
/var/lib/jenkins/jobs/AEB/workspace/app/controllers/services_controller.rb
/var/lib/jenkins/jobs/AEB/workspace/app/controllers/concerns/errors.rb
```

**Comment by hugopeixoto** [comment]
This seems to be caused by `StringFilter#matches?` and `ArrayFilter#matches?`, as they use a regexp to match the whole filename against the given string.

A possible solution could be to replace this:

    (source_file.filename =~ /#{filter_argument}/)

with this:

    fn = source_file.filename.sub(/\A#{Regexp.escape(SimpleCov.root.to_s)}/, "")
    fn  =~ /#{filter_argument}/

or maybe this (with the regexp being precomputed in the initialize method):

    re = /\A#{Regexp.escape(SimpleCov.root.to_s)}.*#{Regexp.escape(filter_argument)}/
    source_file.filename  =~ re

These filters aren't using `Regexp.escape` on the filter_argument, so the second approach might break in some weird cases like `add_group "Libraries", "\Alib"`. The first approach is probably more sturdy.

Any opinions on this?

Both solutions pass the current tests. I could add a test for this and implement the fix using the first approach, if you agree.

**Comment by MaienM** [comment]
Sorry for the late response.

I'd personally use pathname to make the path relative, as it seems (to me) more suited to modify a path than a regex. Something along the lines of

```ruby
fn = Pathname.new(source_file.filename).relative_path_from(SimpleCov.root)
fn ~= /#{filter_argument}/
```

Other than that (which might just boil down to personal preference), that seems great! Thanks you for looking into this.

**Comment by lcguida** [comment]
@MaienM 

This works form me in `ruby 2.2.1`
```ruby
fn = Pathname.new(source_file.filename).relative_path_from(Pathname.new(SimpleCov.root))
fn.to_s ~= /#{filter_argument}/
```

**Comment by PragTob** [comment]
Seemingly fixed in #589 if not please let us know :)
